### Index
1.  [Load & Discover Data](#1)<br>
#### Review Columns and Values
2.  [PurchaseId Column ](#2) <br>
3.  [PurchaseDate - ShipDate Columns](#3) <br>
4.  [ShipMode Column](#4) <br>
5.  [CustomerId Column](#5) <br>
6.  [CustomerName Column](#6) <br>
7.  [Segment Column](#7) <br>
8.  [Country Column](#8) <br>
9.  [City Column](#9) <br>
10. [FirstName-LastName & CustomerId Matches](#10) <br> 
11. [State Column](#11) <br> 
12. [PostalCode Column](#12) <br> 
13. [Region Column](#13) <br> 
14. [ProductId Column](#14) <br> 
15. [Category Column](#15) <br> 

### 1: Load & Discover Data<a id="1"></a>

In [78]:
# import libraries
# import libraries
# Import Required Libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import os
import sqlite3
from sqlite3 import Error

plt.style.use('ggplot')
pd.set_option('display.max_columns', 200) # to display all rows

In [79]:
# Check encoding of dataset

import chardet

# Read the first few bytes to guess the encoding to load the data
with open('Sample-Superstore.csv', 'rb') as file:
    rawdata = file.read(10000)
    result = chardet.detect(rawdata)
    
result

{'encoding': 'ISO-8859-1', 'confidence': 0.73, 'language': ''}

In [80]:
# Load the data using the detected encoding

df = pd.read_csv('Sample-Superstore.csv', sep=',', header=0, encoding='ISO-8859-1').copy()

In [81]:
df.shape

(9994, 21)

In [82]:
df.head().T

0  \
Row ID                                         1   
Order ID                          CA-2016-152156   
Order Date                             11/8/2016   
Ship Date                             11/11/2016   
Ship Mode                           Second Class   
Customer ID                             CG-12520   
Customer Name                        Claire Gute   
Segment                                 Consumer   
Country                            United States   
City                                   Henderson   
State                                   Kentucky   
Postal Code                                42420   
Region                                     South   
Product ID                       FUR-BO-10001798   
Category                               Furniture   
Sub-Category                           Bookcases   
Product Name   Bush Somerset Collection Bookcase   
Sales                                     261.96   
Quantity                                       2   
Discount                                     0.0   
Profit                                   41.9136   

                                                               1  \
Row ID                                                         2   
Order ID                                          CA-2016-152156   
Order Date                                             11/8/2016   
Ship Date                                             11/11/2016   
Ship Mode                                           Second Class   
Customer ID                                             CG-12520   
Customer Name                                        Claire Gute   
Segment                                                 Consumer   
Country                                            United States   
City                                                   Henderson   
State                                                   Kentucky   
Postal Code                                                42420   
Region                                                     South   
Product ID                                       FUR-CH-10000454   
Category                                               Furniture   
Sub-Category                                              Chairs   
Product Name   Hon Deluxe Fabric Upholstered Stacking Chairs,...   
Sales                                                     731.94   
Quantity                                                       3   
Discount                                                     0.0   
Profit                                                   219.582   

                                                               2  \
Row ID                                                         3   
Order ID                                          CA-2016-138688   
Order Date                                             6/12/2016   
Ship Date                                              6/16/2016   
Ship Mode                                           Second Class   
Customer ID                                             DV-13045   
Customer Name                                    Darrin Van Huff   
Segment                                                Corporate   
Country                                            United States   
City                                                 Los Angeles   
State                                                 California   
Postal Code                                                90036   
Region                                                      West   
Product ID                                       OFF-LA-10000240   
Category                                         Office Supplies   
Sub-Category                                              Labels   
Product Name   Self-Adhesive Address Labels for Typewriters b...   
Sales                                                      14.62   
Quantity                                                       2   
Discount                                     

In [83]:
df = df.drop(columns=['Row ID'])

In [84]:
# Rename columns 

df.rename(columns={
    'Order ID': 'PurchaseId',
    'Order Date': 'PurchaseDate',
    'Ship Date': 'ShipDate',
    'Ship Mode': 'ShipMode',  
    'Customer ID': 'CustomerId',
    'Customer Name': 'CustomerName',
    'Segment': 'Segment',
    'Country': 'Country',
    'City': 'City',
    'State': 'State',
    'Postal Code': 'PostalCode',
    'Region': 'Region',
    'Product ID': 'ProductId',
    'Category': 'Category',
    'Sub-Category': 'SubCategory',
    'Product Name': 'ProductName',
    'Sales': 'Sales',
    'Quantity': 'Quantity',
    'Discount': 'Discount',
    'Profit': 'Profit'
}, inplace=True)

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PurchaseId    9994 non-null   object 
 1   PurchaseDate  9994 non-null   object 
 2   ShipDate      9994 non-null   object 
 3   ShipMode      9994 non-null   object 
 4   CustomerId    9994 non-null   object 
 5   CustomerName  9994 non-null   object 
 6   Segment       9994 non-null   object 
 7   Country       9994 non-null   object 
 8   City          9994 non-null   object 
 9   State         9994 non-null   object 
 10  PostalCode    9994 non-null   int64  
 11  Region        9994 non-null   object 
 12  ProductId     9994 non-null   object 
 13  Category      9994 non-null   object 
 14  SubCategory   9994 non-null   object 
 15  ProductName   9994 non-null   object 
 16  Sales         9994 non-null   float64
 17  Quantity      9994 non-null   int64  
 18  Discount      9994 non-null 

In [86]:
df.dtypes

PurchaseId       object
PurchaseDate     object
ShipDate         object
ShipMode         object
CustomerId       object
CustomerName     object
Segment          object
Country          object
City             object
State            object
PostalCode        int64
Region           object
ProductId        object
Category         object
SubCategory      object
ProductName      object
Sales           float64
Quantity          int64
Discount        float64
Profit          float64
dtype: object

In [87]:
# Convert data types for columns 

df['PurchaseId'] = df['PurchaseId'].astype(str)

df['PurchaseDate'] = pd.to_datetime(df['PurchaseDate'])  

df['ShipDate'] = pd.to_datetime(df['ShipDate'])  

df['ShipMode'] = df['ShipMode'].astype(str)

df['CustomerId'] = df['CustomerId'].astype(str)

df['CustomerName'] = df['CustomerName'].astype(str)

df['Segment'] = df['Segment'].astype(str)

df['Country'] = df['Country'].astype(str)

df['City'] = df['City'].astype(str)

df['State'] = df['State'].astype(str)

df['PostalCode'] = df['PostalCode'].astype(str)  

df['Region'] = df['Region'].astype(str)

df['ProductId'] = df['ProductId'].astype(str)

df['Category'] = df['Category'].astype(str)

df['SubCategory'] = df['SubCategory'].astype(str)

df['ProductName'] = df['ProductName'].astype(str)

df['Sales'] = df['Sales'].astype(float)  

df['Quantity'] = df['Quantity'].astype(int)

df['Discount'] = df['Discount'].astype(float)

df['Profit'] = df['Profit'].astype(float)

In [88]:
df.isna().sum()

PurchaseId      0
PurchaseDate    0
ShipDate        0
ShipMode        0
CustomerId      0
CustomerName    0
Segment         0
Country         0
City            0
State           0
PostalCode      0
Region          0
ProductId       0
Category        0
SubCategory     0
ProductName     0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

In [89]:
df.nunique()

PurchaseId      5009
PurchaseDate    1237
ShipDate        1334
ShipMode           4
CustomerId       793
CustomerName     793
Segment            3
Country            1
City             531
State             49
PostalCode       631
Region             4
ProductId       1862
Category           3
SubCategory       17
ProductName     1850
Sales           5825
Quantity          14
Discount          12
Profit          7287
dtype: int64

In [90]:
df.columns

Index(['PurchaseId', 'PurchaseDate', 'ShipDate', 'ShipMode', 'CustomerId',
       'CustomerName', 'Segment', 'Country', 'City', 'State', 'PostalCode',
       'Region', 'ProductId', 'Category', 'SubCategory', 'ProductName',
       'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

### 2: PurchaseId<a id="2"></a>

In [91]:
# check if entries have the same format like: CA-2016-152156

import re

pattern = r'^[A-Z]{2}-\d{4}-\d{6}$'

df[~df['PurchaseId'].str.match(pattern)]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

In [92]:
# Trim whitespaces

df['PurchaseId'] = df['PurchaseId'].str.strip()

In [93]:
# Ensure that the case is uppercase across all entries.

df['PurchaseId'] = df['PurchaseId'].str.upper()

In [94]:
# Check if there is any unexpected characters or symbols 

df[df['PurchaseId'].str.contains(r'[^A-Z0-9-]')]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

In [95]:
# Check if the year starts with 20 and followed with same pattern 

df[~df['PurchaseId'].str.contains(r'-20\d{2}-')]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

### 3: PurchaseDate - ShipDate<a id="3"></a>

In [96]:
# ensure that the PurchaseDate is always earlier than the ShipDate 

df[df['PurchaseDate'] > df['ShipDate']]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

In [97]:
# check entries where OrderDate and ShipDate is same

df_order_ship_same = df[df['PurchaseDate'] == df['ShipDate']]

df_order_ship_same.head()

PurchaseId PurchaseDate   ShipDate  ShipMode CustomerId  \
366  CA-2016-155516   2016-10-21 2016-10-21  Same Day   MK-17905   
367  CA-2016-155516   2016-10-21 2016-10-21  Same Day   MK-17905   
368  CA-2016-155516   2016-10-21 2016-10-21  Same Day   MK-17905   
369  CA-2016-155516   2016-10-21 2016-10-21  Same Day   MK-17905   
657  US-2016-156097   2016-09-19 2016-09-19  Same Day   EH-14125   

          CustomerName      Segment        Country        City        State  \
366    Michael Kennedy    Corporate  United States  Manchester  Connecticut   
367    Michael Kennedy    Corporate  United States  Manchester  Connecticut   
368    Michael Kennedy    Corporate  United States  Manchester  Connecticut   
369    Michael Kennedy    Corporate  United States  Manchester  Connecticut   
657  Eugene Hildebrand  Home Office  United States      Aurora     Illinois   

    PostalCode   Region        ProductId         Category SubCategory  \
366       6040     East  OFF-BI-10002412  Office Supplies     Binders   
367       6040     East  OFF-SU-10001225  Office Supplies    Supplies   
368       6040     East  OFF-ST-10002406  Office Supplies     Storage   
369       6040     East  FUR-BO-10002545        Furniture   Bookcases   
657      60505  Central  FUR-CH-10001215        Furniture      Chairs   

                                           ProductName     Sales  Quantity  \
366  Wilson Jones “Snap” Scratch Pad Binder Tool fo...    23.200         4   
367                                     Staple remover     7.360         2   
368                           Pizazz Global Quick File   104.790         7   
369  Atlantic Metals Mobile 3-Shelf Bookcases, Cust...  1043.920         4   
657      Global Troy Executive Leather Low-Back Tilter   701.372         2   

     Discount    Profit  
366       0.0   10.4400  
367       0.0    0.1472  
368       0.0   29.3412  
369       0.0  271.4192  
657       0.3  -50.0980

In [98]:
df_order_ship_same.info()

<class 'pandas.core.frame.DataFrame'>
Index: 519 entries, 366 to 9963
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   PurchaseId    519 non-null    object        
 1   PurchaseDate  519 non-null    datetime64[ns]
 2   ShipDate      519 non-null    datetime64[ns]
 3   ShipMode      519 non-null    object        
 4   CustomerId    519 non-null    object        
 5   CustomerName  519 non-null    object        
 6   Segment       519 non-null    object        
 7   Country       519 non-null    object        
 8   City          519 non-null    object        
 9   State         519 non-null    object        
 10  PostalCode    519 non-null    object        
 11  Region        519 non-null    object        
 12  ProductId     519 non-null    object        
 13  Category      519 non-null    object        
 14  SubCategory   519 non-null    object        
 15  ProductName   519 non-null    object      

#### 519 rows have the same Order Date and Ship Date which are the Same Day shipment. 
#### No cases of Ship Date occurring before Order Date.
#### All entries in these columns are consistent.

## 4: ShipMode<a id="4"></a>

In [99]:
# check unique values in ShipMode column

df['ShipMode'].unique()

array(['Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [100]:
# Ensure that the values are consistently spelled and capitalized.

df['ShipMode'] = df['ShipMode'].str.strip().str.title()

In [101]:
# Review distribution of 'ShipMode' values

df['ShipMode'].value_counts()

ShipMode
Standard Class    5968
Second Class      1945
First Class       1538
Same Day           543
Name: count, dtype: int64

## 5: CustomerId<a id="5"></a>

In [102]:
# Make sure all CustomerId values follow a consistent format 

valid_id_pattern = r'^[A-Z]{2}-\d{5}$'

df[~df['CustomerId'].str.match(valid_id_pattern)].T

261   \
PurchaseId                                       US-2017-155299   
PurchaseDate                                2017-06-08 00:00:00   
ShipDate                                    2017-06-12 00:00:00   
ShipMode                                         Standard Class   
CustomerId                                             Dl-13600   
CustomerName                                       Dorris liebe   
Segment                                               Corporate   
Country                                           United States   
City                                                   Pasadena   
State                                                     Texas   
PostalCode                                                77506   
Region                                                  Central   
ProductId                                       OFF-AP-10002203   
Category                                        Office Supplies   
SubCategory                                          Appliances   
ProductName   Eureka Disposable Bags for Sanitaire Vibra Gro...   
Sales                                                     1.624   
Quantity                                                      2   
Discount                                                    0.8   
Profit                                                   -4.466   

                                                    701   \
PurchaseId                                CA-2017-114552   
PurchaseDate                         2017-09-02 00:00:00   
ShipDate                             2017-09-08 00:00:00   
ShipMode                                  Standard Class   
CustomerId                                      Dl-13600   
CustomerName                                Dorris liebe   
Segment                                        Corporate   
Country                                    United States   
City                                           Cleveland   
State                                               Ohio   
PostalCode                                         44105   
Region                                              East   
ProductId                                FUR-FU-10002960   
Category                                       Furniture   
SubCategory                                  Furnishings   
ProductName   Eldon 200 Class Desk Accessories, Burgundy   
Sales                                             15.072   
Quantity                                               3   
Discount                                             0.2   
Profit                                            4.1448   

                                               715                  1489  \
PurchaseId                           CA-2014-153150       CA-2014-136280   
PurchaseDate                    2014-07-01 00:00:00  2014-11-29 00:00:00   
ShipDate                        2014-07-06 00:00:00  2014-12-06 00:00:00   
ShipMode                               Second Class       Standard Class   
CustomerId                                 Dl-13600             Co-12640   
CustomerName                           Dorris liebe           Corey-Lock   
Segment                                   Corporate             Consumer   
Country                               United States        United States   
City                                        Seattle         Philadelphia   
State                                    Washington         Pennsylvania   
PostalCode                                    98105                19143   
Region                                         West                 East   
ProductId                           OFF-BI-10003355      OFF-LA-10000452   
Category                            Office Supplies      Office Supplies   
SubCategory                                 Binders               Labels   
ProductName   Cardinal Holdit Business Card Pockets            Avery 488   
Sales                                         19.92                 5.04   
Quantity              

#### There seem to be upper-lower case issues for CustomerId.

In [103]:
# Convert all 'CustomerId' values to uppercase

df['CustomerId'] = df['CustomerId'].str.upper()

In [104]:
# Confirm Changes

df[~df['CustomerId'].str.match(valid_id_pattern)]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

In [105]:
# Clean probable leading or trailing spaces.

df['CustomerId'] = df['CustomerId'].str.strip()

### All values in CustomerId column have the same format now

## 6: CustomerName<a id="6"></a>

In [106]:
# Trim whitespaces from 'CustomerName' values
df['CustomerName'] = df['CustomerName'].str.strip()

# Convert 'CustomerName' values to title case
df['CustomerName'] = df['CustomerName'].str.title()

#### Noticed some names which have hypnes in between such as: Corey-Lock

In [107]:
# Noticed some names which have hypnes in between such as: Corey-Lock
# Regular expression pattern to detect hyphenated names

hyphenated_pattern = r'\b\w+-\w+\b'

# Function to check if a name contains a hyphen

def contains_hyphen(name):
    return bool(re.search(hyphenated_pattern, name))

# Apply the function to get a boolean Series

mask1 = df['CustomerName'].apply(contains_hyphen)

In [108]:
df['CustomerName'][mask1]

1489    Corey-Lock
1900    Corey-Lock
2966    Corey-Lock
2967    Corey-Lock
4643    Corey-Lock
4644    Corey-Lock
8145    Corey-Lock
8146    Corey-Lock
Name: CustomerName, dtype: object

It is the only Name with hypnes between names. 

In [109]:
# Correct the specific entry "Corey-Lock" to "Corey Lock"

df.loc[df['CustomerName'] == 'Corey-Lock', 'CustomerName'] = 'Corey Lock'

In [110]:
# Also noticed some name with apostrophes, filter and review values to be sure that there is no issue 
# Regular expression pattern to detect names with apostrophes 
apostrophe_pattern = r"\b\w*'\w*\b"

# Function to check if a name contains an apostrophe
def contains_apostrophe(name):
    return bool(re.search(apostrophe_pattern, name))

# Apply the function to create a boolean mask
mask2 = df['CustomerName'].apply(contains_apostrophe)

df[mask2].T

3     \
PurchaseId                                   US-2015-108966   
PurchaseDate                            2015-10-11 00:00:00   
ShipDate                                2015-10-18 00:00:00   
ShipMode                                     Standard Class   
CustomerId                                         SO-20335   
CustomerName                                 Sean O'Donnell   
Segment                                            Consumer   
Country                                       United States   
City                                        Fort Lauderdale   
State                                               Florida   
PostalCode                                            33311   
Region                                                South   
ProductId                                   FUR-TA-10000577   
Category                                          Furniture   
SubCategory                                          Tables   
ProductName   Bretford CR4500 Series Slim Rectangular Table   
Sales                                              957.5775   
Quantity                                                  5   
Discount                                               0.45   
Profit                                             -383.031   

                                        4     \
PurchaseId                    US-2015-108966   
PurchaseDate             2015-10-11 00:00:00   
ShipDate                 2015-10-18 00:00:00   
ShipMode                      Standard Class   
CustomerId                          SO-20335   
CustomerName                  Sean O'Donnell   
Segment                             Consumer   
Country                        United States   
City                         Fort Lauderdale   
State                                Florida   
PostalCode                             33311   
Region                                 South   
ProductId                    OFF-ST-10000760   
Category                     Office Supplies   
SubCategory                          Storage   
ProductName   Eldon Fold 'N Roll Cart System   
Sales                                 22.368   
Quantity                                   2   
Discount                                 0.2   
Profit                                2.5164   

                                                           46    \
PurchaseId                                       CA-2014-146703   
PurchaseDate                                2014-10-20 00:00:00   
ShipDate                                    2014-10-25 00:00:00   
ShipMode                                           Second Class   
CustomerId                                             PO-18865   
CustomerName                                  Patrick O'Donnell   
Segment                                                Consumer   
Country                                           United States   
City                                                   Westland   
State                                                  Michigan   
PostalCode                                                48185   
Region                                                  Central   
ProductId                                       OFF-ST-10001713   
Category                                        Office Supplies   
SubCategory                                             Storage   
ProductName   Gould Plastics 9-Pocket Panel Bin, 18-3/8w x 5...   
Sales                                                    211.96   
Quantity                                                      4   
Discount                                                    0.0   
Profit                                                   8.4784   

                                          85    \
PurchaseId                      CA-2017-140088   
PurchaseDate               2017-05-28 00:00:00   
ShipDate                   2017-05-30 00:00:00   
ShipMode                          Second Class   
CustomerId                            PO-18865   
CustomerName  

In [111]:
# Ensure there is space between name and surname 

# Regular expression pattern to detect names with at least one space 
space_pattern = r'\s'

# Function to check if a name contains a space
def contains_space(name):
    return bool(re.search(space_pattern, name))

# Apply the function to create a boolean mask
mask3 = df['CustomerName'].apply(contains_space)

# Filter the DataFrame to find names without spaces
df[~mask3]

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit]
Index: []

In [112]:
# Split the CustomerName column based on the last space and create 2 columns as: FirstName, LastName

df[['FirstName', 'LastName']] = df['CustomerName'].str.rsplit(n=1, expand=True)

In [113]:
# Check the frequency distribution of names to identify common and rare names.

# Get frequency count of each name

first_name_counts = df['FirstName'].value_counts()

 # Display most common names
 
first_name_counts.tail(20)

FirstName
Grant         6
Angele        5
Georgia       5
Claire        5
Aleksandra    5
Melanie       5
Astrea        5
Victor        5
Elpida        5
Jenna         4
Eileen        4
Jasper        4
Ionia         4
Ritsa         3
Thais         2
Patricia      2
Stefanie      2
Anemone       2
Jocasta       1
Lela          1
Name: count, dtype: int64

In [114]:
# Get frequency count of each lastname

last_name_counts = df['LastName'].value_counts()

 # Display least common lastnames

last_name_counts.tail(20)

LastName
Caffey       4
Taslimi      3
Weirich      3
Skaria       3
Hightower    3
Jenkins      3
Shami        3
Tron         3
Thomas       3
Hale         3
Blacks       3
Ratner       2
Holloman     2
Hirasaki     2
Sissman      2
Odegard      2
Breyer       2
Donovan      1
Rupert       1
Emerson      1
Name: count, dtype: int64

In [115]:
# Function to check for unusual characters and patterns

def is_unusual_name(name):
    # Check for names with digits
    if re.search(r'[0-9]', name):  # Names with digits
        return True
    # Check for names with invalid special characters, excluding apostrophes and 'ö'
    if re.search(r'[^a-zA-Z\s\'ö-]', name):  # Names with invalid special characters
        return True
    # Check for unusually short or long names
    if len(name) < 2 or len(name) > 50:  # Unusually short or long names
        return True
    return False

# Apply function to identify unusual names

df['UnusualName'] = df['FirstName'].apply(is_unusual_name)

# Filter names flagged as unusual

unusual_names_df = df[df['UnusualName']]

unusual_names_df

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, UnusualName]
Index: []

In [116]:
# Function to check for unusual characters and patterns
# included 'ö', 'ä', 'ü' as we have Eurepean names which have these characters 

def is_unusual_lastname(name):
    # Check for names with digits
    if re.search(r'[0-9]', name):
        return True
    # Check for names with invalid special characters, excluding apostrophes, 'ö', 'ä', 'ü', and hyphens
    if re.search(r'[^a-zA-Z\s\'öäü-]', name):  # Names with invalid special characters
        return True
    # Check for unusually short or long names
    if len(name) < 2 or len(name) > 50:  # Unusually short or long names
        return True
    return False

# Apply function to identify unusual last names

df['UnusualLastName'] = df['LastName'].apply(is_unusual_lastname)

# Filter names flagged as unusual

unusual_lastnames_df = df[df['UnusualLastName']]

# Display the DataFrame with unusual last names

unusual_lastnames_df.T

912   \
PurchaseId                         CA-2015-133627   
PurchaseDate                  2015-05-31 00:00:00   
ShipDate                      2015-06-07 00:00:00   
ShipMode                           Standard Class   
CustomerId                               SC-20050   
CustomerName                     Sample Company A   
Segment                               Home Office   
Country                             United States   
City                                      Norwich   
State                                 Connecticut   
PostalCode                                   6360   
Region                                       East   
ProductId                         FUR-FU-10001935   
Category                                Furniture   
SubCategory                           Furnishings   
ProductName      3M Hangers With Command Adhesive   
Sales                                        22.2   
Quantity                                        6   
Discount                                      0.0   
Profit                                      9.102   
FirstName                          Sample Company   
LastName                                        A   
UnusualName                                 False   
UnusualLastName                              True   

                                                              2004  \
PurchaseId                                          US-2017-143028   
PurchaseDate                                   2017-04-11 00:00:00   
ShipDate                                       2017-04-18 00:00:00   
ShipMode                                            Standard Class   
CustomerId                                                SC-20050   
CustomerName                                      Sample Company A   
Segment                                                Home Office   
Country                                              United States   
City                                                       Lubbock   
State                                                        Texas   
PostalCode                                                   79424   
Region                                                     Central   
ProductId                                          OFF-BI-10004738   
Category                                           Office Supplies   
SubCategory                                                Binders   
ProductName      Flexible Leather- Look Classic Collection Ring...   
Sales                                                       11.364   
Quantity                                                         3   
Discount                                                       0.8   
Profit                                                     -17.046   
FirstName                                           Sample Company   
LastName                                                         A   
UnusualName                                                  False   
UnusualLastName                                               True   

                                             3661  \
PurchaseId                         CA-2016-155005   
PurchaseDate                  2016-06-13 00:00:00   
ShipDate                      2016-06-15 00:00:00   
ShipMode                             Second Class   
CustomerId                               SC-20050   
CustomerName                     Sample Company A   
Segment                               Home Office   
Country                             United States   
City                                      Jackson   
State                                    Michigan   
PostalCode                                  49201   
Region                                    Central   
ProductId                         TEC-PH-10003484   
Category                               Technology   
SubCategory                                Phones   
ProductName      Ooma Telo VoIP Home Phone System   
Sales                                      377.97   
Quantity      

In [117]:
unusual_lastnames_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15 entries, 912 to 9771
Data columns (total 24 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   PurchaseId       15 non-null     object        
 1   PurchaseDate     15 non-null     datetime64[ns]
 2   ShipDate         15 non-null     datetime64[ns]
 3   ShipMode         15 non-null     object        
 4   CustomerId       15 non-null     object        
 5   CustomerName     15 non-null     object        
 6   Segment          15 non-null     object        
 7   Country          15 non-null     object        
 8   City             15 non-null     object        
 9   State            15 non-null     object        
 10  PostalCode       15 non-null     object        
 11  Region           15 non-null     object        
 12  ProductId        15 non-null     object        
 13  Category         15 non-null     object        
 14  SubCategory      15 non-null     object      

In [118]:
# Identify and correct Sample Company entry in FirstName', 'LastName columns

df.loc[(df['FirstName'] == 'Sample Company') & (df['LastName'] == 'A'), ['FirstName', 'LastName']] = ['Sample', 'Company']

In [119]:
# Ensure that values are updated in CustomerName, FirstName', 'LastName columns


# Apply function to identify unusual last names

df['UnusualLastName'] = df['LastName'].apply(is_unusual_lastname)

# Filter names flagged as unusual

unusual_lastnames_df = df[df['UnusualLastName']]

# Display the DataFrame with unusual last names

unusual_lastnames_df

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, UnusualName, UnusualLastName]
Index: []

In [120]:
# Sort the DataFrame alphabetically by 'FirstName' to check some values in FirstName column

sorted_df = df.sort_values(by='FirstName', ascending=True)

sorted_df['FirstName'].unique()

array(['Aaron', 'Adam', 'Adrian', 'Aimee', 'Alan', 'Alejandro',
       'Aleksandra', 'Alex', 'Alice', 'Allen', 'Alyssa', 'Amy', 'Andrew',
       'Andy', 'Anemone', 'Angele', 'Ann', 'Anna', 'Anne', 'Annie',
       'Anthony', 'Arianne', 'Art', 'Arthur', 'Ashley', 'Astrea',
       'Barbara', 'Barry', 'Bart', 'Becky', 'Ben', 'Benjamin', 'Berenike',
       'Beth', 'Bill', 'Bobby', 'Brad', 'Bradley', 'Brenda', 'Brendan',
       'Brian', 'Brooke', 'Brosina', 'Bruce', 'Bryan', 'Candace', 'Cari',
       'Carl', 'Carlos', 'Carol', 'Caroline', 'Cassandra', 'Catherine',
       'Cathy', 'Chad', 'Charles', 'Charlotte', 'Chloris', 'Chris',
       'Christina', 'Christine', 'Christopher', 'Christy', 'Chuck',
       'Cindy', 'Claire', 'Claudia', 'Clay', 'Clytie', 'Corey', 'Corinna',
       'Craig', 'Cyma', 'Cynthia', 'Cyra', 'Damala', 'Dan', 'Dana',
       'Daniel', 'Dario', 'Darren', 'Darrin', 'Darrin Van', 'Dave',
       'David', 'Dean', 'Deanra', 'Deborah', 'Debra', 'Deirdre',
       'Delfina', 'Deni

In [121]:
# Some Entries has Zuschuss as FirstName value which does not seem like a real name. 

# Step 1: Create a boolean mask where 'FirstName' is 'Zuschuss'
df_name_test = df['FirstName'] == 'Zuschuss'

# Step 2: Apply the mask to filter the DataFrame
filtered_df = df[df_name_test]

filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 40 entries, 18 to 8923
Data columns (total 24 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   PurchaseId       40 non-null     object        
 1   PurchaseDate     40 non-null     datetime64[ns]
 2   ShipDate         40 non-null     datetime64[ns]
 3   ShipMode         40 non-null     object        
 4   CustomerId       40 non-null     object        
 5   CustomerName     40 non-null     object        
 6   Segment          40 non-null     object        
 7   Country          40 non-null     object        
 8   City             40 non-null     object        
 9   State            40 non-null     object        
 10  PostalCode       40 non-null     object        
 11  Region           40 non-null     object        
 12  ProductId        40 non-null     object        
 13  Category         40 non-null     object        
 14  SubCategory      40 non-null     object       

#### Entries with 'Zuschuss' as FirstName:

    There are 40 records where the FirstName is 'Zuschuss'. These entries belong to 2 distinct customers.
    We will not be removing or editing these records at this time, as we can not communicate and solve it with business owner. However, this has been noted.

#### Entries with 'Sample Company' and 'A':

    There are 15 records where the CustomerName is 'Sample Company' and the LastName is 'A'.
    We have decided not to remove these records as they contain valuable information in other fields. This issue has also been noted for future correction.

#### The FirstName, LastName, and CustomerName columns have been reviewed to ensure they have the correct format and do not contain unusual values.
#### 8 Entry were corrected which include an invalid character(hypens) (1 Customer with 8 entries: Corey-Lock)

## 7: Segment<a id="7"></a>

In [122]:
df['Segment'].unique()

array(['Consumer', 'Corporate', 'Home Office'], dtype=object)

In [123]:
df['Segment'].value_counts()

Segment
Consumer       5191
Corporate      3020
Home Office    1783
Name: count, dtype: int64

## 8: Country<a id="8"></a>

In [124]:
df['Country'].nunique()

1

In [125]:
df['Country'].unique()

array(['United States'], dtype=object)

## 9: City<a id="9"></a>

In [126]:
df.nunique()

PurchaseId         5009
PurchaseDate       1237
ShipDate           1334
ShipMode              4
CustomerId          793
CustomerName        793
Segment               3
Country               1
City                531
State                49
PostalCode          631
Region                4
ProductId          1862
Category              3
SubCategory          17
ProductName        1850
Sales              5825
Quantity             14
Discount             12
Profit             7287
FirstName           340
LastName            588
UnusualName           1
UnusualLastName       1
dtype: int64

In [127]:
# Create a new DataFrame that shows the count of each unique city in the City column

city_counts = df['City'].value_counts()

city_counts_df = city_counts.reset_index()
city_counts_df.columns = ['City', 'Count']

city_counts_df = city_counts_df.sort_values(by='Count', ascending=False)

city_counts_df

City  Count
0        New York City    915
1          Los Angeles    747
2         Philadelphia    537
3        San Francisco    510
4              Seattle    428
..                 ...    ...
483         Hagerstown      1
482  Arlington Heights      1
481            Baytown      1
479    San Luis Obispo      1
530          Manhattan      1

[531 rows x 2 columns]

In [128]:
city_counts_df.head(50).T

0            1             2              3        4   \
City   New York City  Los Angeles  Philadelphia  San Francisco  Seattle   
Count            915          747           537            510      428   

            5        6         7          8            9       10  \
City   Houston  Chicago  Columbus  San Diego  Springfield  Dallas   
Count      377      314       222        170          163     157   

                 11       12      13        14       15        16      17  \
City   Jacksonville  Detroit  Newark  Richmond  Jackson  Columbia  Aurora   
Count           125      115      95        90       82        81      68   

            18          19         20           21     22          23  \
City   Phoenix  Long Beach  Arlington  San Antonio  Miami  Louisville   
Count       63          61         60           59     57          57   

              24         25         26        27         28         29  \
City   Rochester  Charlotte  Henderson  Lakewood  Lancaster  Fairfield   
Count         53         52         51        49         46         45   

              30      31        32         33        35        36         34  \
City   Milwaukee  Denver  Lawrence  Baltimore  Pasadena  San Jose  Cleveland   
Count         45      44        44         43        42        42         42   

                 37     38       39      40        41     42          43  \
City   Fayetteville  Salem  Atlanta  Austin  Franklin  Tampa  Wilmington   
Count            41     40       39      39        37     36          36   

               44       45      46      47          48         49  
City   Huntsville  Decatur  Toledo  Tucson  Providence  Lafayette  
Count          36       35      32      32          31         31

In [129]:
city_counts_df.tail(50).T

496            527       526        525            524  \
City   Redwood City  Missouri City  Glenview  San Mateo  Commerce City   
Count             1              1         1          1              1   

           522     515        521         520       519          518     517  \
City   Holyoke  Normal  Goldsboro  Montebello  Waukesha  Orland Park  Conroe   
Count        1       1          1           1         1            1       1   

           516       497        480        495          468      474      473  \
City   Abilene  Bartlett  Champaign  La Quinta  Tinley Park  Antioch  Ontario   
Count        1         1          1          1            1        1        1   

             472            471      470    469        467           476  \
City   Melbourne  Atlantic City  Portage  Davis  Littleton  Ormond Beach   
Count          1              1        1      1          1             1   

                  466     465     464      463        462        461  \
City   Citrus Heights  Linden  Rogers  Danbury  Iowa City  Kissimmee   
Count               1       1       1        1          1          1   

             475             477     494        487          493      492  \
City   Deer Park  Jefferson City  Conway  Rock Hill  Port Orange  Margate   
Count          1               1       1          1            1        1   

            491            490         489       488           486  \
City   Missoula  Lake Elsinore  Romeoville  Murrieta  Saint Peters   
Count         1              1           1         1             1   

              478       485     484         483                482      481  \
City   Springdale  Aberdeen  Elyria  Hagerstown  Arlington Heights  Baytown   
Count           1         1       1           1                  1        1   

                   479        530  
City   San Luis Obispo  Manhattan  
Count                1          1

In [130]:
city_counts_df = city_counts_df.sort_values(by='City')

city_counts_df.head(50).T

485      516    80           118         97     335        212  \
City   Aberdeen  Abilene  Akron  Albuquerque  Alexandria  Allen  Allentown   
Count         1        1     21           14          16      4          7   

           407       167      58       345        272      474     210  \
City   Altoona  Amarillo  Anaheim  Andover  Ann Arbor  Antioch  Apopka   
Count        2        10       27        4          5        1       7   

                178       399        20                 482     340  \
City   Apple Valley  Appleton  Arlington  Arlington Heights  Arvada   
Count             9         2         60                  1       4   

             225     192      39             471     70      17      40   \
City   Asheville  Athens  Atlanta  Atlantic City  Auburn  Aurora  Austin   
Count          7       8       39              1      24      68      39   

            264          99         33      266       497      383      481  \
City   Avondale  Bakersfield  Baltimore  Bangor  Bartlett  Bayonne  Baytown   
Count         6           16         43       5         1        3        1   

            280      269         166       251         386        268  \
City   Beaumont  Bedford  Belleville  Bellevue  Bellingham  Bethlehem   
Count         5        5          10         6           3          5   

           369       498          106         353    320          259  \
City   Beverly  Billings  Bloomington  Boca Raton  Boise  Bolingbrook   
Count        3         1           15           3      4            6   

                241            152            154      424        103  
City   Bossier City  Bowling Green  Boynton Beach  Bozeman  Brentwood  
Count             6             10             10        2         16

In [131]:
city_counts_df.tail(50).T

158            278     206          468     46        315  \
City   Thornton  Thousand Oaks  Tigard  Tinley Park  Toledo  Torrance   
Count        10              5       8            1      32         4   

           125   55      47     63          451         432    284        357  \
City   Trenton  Troy  Tucson  Tulsa  Tuscaloosa  Twin Falls  Tyler  Urbandale   
Count       13    29      32     26           2           2      5          3   

         195        509      252        293       203             98   \
City   Utica  Vacaville  Vallejo  Vancouver  Vineland  Virginia Beach   
Count      8          1        6          5         8              16   

           321   257            440      294         153        147       514  \
City   Visalia  Waco  Warner Robins  Warwick  Washington  Waterbury  Waterloo   
Count        4     6              2        5          10         11         1   

             162       519     309         164         444          300  \
City   Watertown  Waukesha  Wausau  Waynesboro  West Allis  West Jordan   
Count         10         1       4          10           2            5   

                   356        238       122          93        215       513  \
City   West Palm Beach  Westfield  Westland  Westminster  Wheeling  Whittier   
Count                3          6        13           17         7         1   

           218         43      233       358       382        275         305  \
City   Wichita  Wilmington  Wilson  Woodbury  Woodland  Woodstock  Woonsocket   
Count        7          36       6         3         3          5           4   

           108   296      523   311  
City   Yonkers  York  Yucaipa  Yuma  
Count       15     5        1     4

In [132]:
df['City'] = df['City'].str.strip()

df['City'] = df['City'].str.replace(r'\s+', ' ', regex=True)

In [133]:
df['City'].nunique()

531

## 10: FirstName-LastName & CustomerId Matches<a id="10"></a>

### First Letter of Customer Name and Surname should be the first 2 letters of CustomerId according to dataset 

In [134]:
# Define a function to validate the CustomerId

def validate_customer_id(row):
    expected_prefix = row['FirstName'][0] + row['LastName'][0]
    return row['CustomerId'].startswith(expected_prefix)

df['ValidCustomerId'] = df.apply(validate_customer_id, axis=1)

# Filter rows where CustomerId does not follow the rule
invalid_customer_ids_df = df[~df['ValidCustomerId']]

invalid_customer_ids_df.head()

PurchaseId PurchaseDate   ShipDate        ShipMode CustomerId  \
2    CA-2016-138688   2016-06-12 2016-06-16    Second Class   DV-13045   
487  CA-2014-154627   2014-10-29 2014-10-31     First Class   SA-20830   
788  CA-2015-115938   2015-06-26 2015-06-30  Standard Class   SA-20830   
789  CA-2015-115938   2015-06-26 2015-06-30  Standard Class   SA-20830   
790  CA-2015-115938   2015-06-26 2015-06-30  Standard Class   SA-20830   

        CustomerName    Segment        Country         City       State  \
2    Darrin Van Huff  Corporate  United States  Los Angeles  California   
487     Sue Ann Reed   Consumer  United States      Chicago    Illinois   
788     Sue Ann Reed   Consumer  United States     Richmond    Virginia   
789     Sue Ann Reed   Consumer  United States     Richmond    Virginia   
790     Sue Ann Reed   Consumer  United States     Richmond    Virginia   

    PostalCode   Region        ProductId         Category SubCategory  \
2        90036     West  OFF-LA-10000240  Office Supplies      Labels   
487      60610  Central  TEC-PH-10001363       Technology      Phones   
788      23223    South  OFF-BI-10001543  Office Supplies     Binders   
789      23223    South  OFF-ST-10001321  Office Supplies     Storage   
790      23223    South  OFF-BI-10001132  Office Supplies     Binders   

                                           ProductName     Sales  Quantity  \
2    Self-Adhesive Address Labels for Typewriters b...    14.620         2   
487                                    Apple iPhone 5S  2735.952         6   
788               GBC VeloBinder Manual Binding System   143.960         4   
789        Decoflex Hanging Personal Folder File, Blue    15.420         1   
790  Acco PRESSTEX Data Binder with Storage Hooks, ...    43.040         8   

     Discount    Profit   FirstName LastName  UnusualName  UnusualLastName  \
2         0.0    6.8714  Darrin Van     Huff        False            False   
487       0.2  341.9940     Sue Ann     Reed        False            False   
788       0.0   69.1008     Sue Ann     Reed        False            False   
789       0.0    4.1634     Sue Ann     Reed        False            False   
790       0.0   21.0896     Sue Ann     Reed        False            False   

     ValidCustomerId  
2              False  
487            False  
788            False  
789            False  
790            False

In [135]:
invalid_customer_ids_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66 entries, 2 to 9018
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   PurchaseId       66 non-null     object        
 1   PurchaseDate     66 non-null     datetime64[ns]
 2   ShipDate         66 non-null     datetime64[ns]
 3   ShipMode         66 non-null     object        
 4   CustomerId       66 non-null     object        
 5   CustomerName     66 non-null     object        
 6   Segment          66 non-null     object        
 7   Country          66 non-null     object        
 8   City             66 non-null     object        
 9   State            66 non-null     object        
 10  PostalCode       66 non-null     object        
 11  Region           66 non-null     object        
 12  ProductId        66 non-null     object        
 13  Category         66 non-null     object        
 14  SubCategory      66 non-null     object        

In [136]:
invalid_customer_ids_df['CustomerName'].value_counts()

CustomerName
Sue Ann Reed       21
Mark Van Huff      15
Paul Van Hugh      13
Darrin Van Huff     9
Corey Lock          8
Name: count, dtype: int64

In [137]:
# Correct the CustomerId values based on CustomerName

df.loc[df['CustomerName'] == 'Corey Lock', 'CustomerId'] = 'CL-12640'
df.loc[df['CustomerName'] == 'Sue Ann Reed', 'CustomerId'] = 'SR-20830'

In [138]:
# Darrin Van Huff
df.loc[df['CustomerName'] == 'Darrin Van Huff', ['FirstName', 'LastName']] = ['Darrin', 'Van Huff']

# Mark Van Huff
df.loc[df['CustomerName'] == 'Mark Van Huff', ['FirstName', 'LastName']] = ['Mark', 'Van Huff']

# Paul Van Hugh
df.loc[df['CustomerName'] == 'Paul Van Hugh', ['FirstName', 'LastName']] = ['Paul', 'Van Hugh']

In [139]:
# confirm changes

df['ValidCustomerId'] = df.apply(validate_customer_id, axis=1)

invalid_customer_ids_df = df[~df['ValidCustomerId']]

invalid_customer_ids_df

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, UnusualName, UnusualLastName, ValidCustomerId]
Index: []

### 66 records from 5 Customer were updated. Listed below: 
* Sue Ann Reed       21
* Mark Van Huff      15 
* Paul Van Hugh      13
* Darrin Van Huff     9
* Corey Lock          8

In [140]:
df.columns

Index(['PurchaseId', 'PurchaseDate', 'ShipDate', 'ShipMode', 'CustomerId',
       'CustomerName', 'Segment', 'Country', 'City', 'State', 'PostalCode',
       'Region', 'ProductId', 'Category', 'SubCategory', 'ProductName',
       'Sales', 'Quantity', 'Discount', 'Profit', 'FirstName', 'LastName',
       'UnusualName', 'UnusualLastName', 'ValidCustomerId'],
      dtype='object')

In [141]:
df = df.drop(columns=['UnusualName', 'UnusualLastName', 'ValidCustomerId'])

df = df.reset_index(drop=True).copy()

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 22 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   PurchaseId    9994 non-null   object        
 1   PurchaseDate  9994 non-null   datetime64[ns]
 2   ShipDate      9994 non-null   datetime64[ns]
 3   ShipMode      9994 non-null   object        
 4   CustomerId    9994 non-null   object        
 5   CustomerName  9994 non-null   object        
 6   Segment       9994 non-null   object        
 7   Country       9994 non-null   object        
 8   City          9994 non-null   object        
 9   State         9994 non-null   object        
 10  PostalCode    9994 non-null   object        
 11  Region        9994 non-null   object        
 12  ProductId     9994 non-null   object        
 13  Category      9994 non-null   object        
 14  SubCategory   9994 non-null   object        
 15  ProductName   9994 non-null   object  

## 11: State<a id="11"></a>

In [142]:
df['State'].nunique()

49

In [143]:
df['State'].unique()

array(['Kentucky', 'California', 'Florida', 'North Carolina',
       'Washington', 'Texas', 'Wisconsin', 'Utah', 'Nebraska',
       'Pennsylvania', 'Illinois', 'Minnesota', 'Michigan', 'Delaware',
       'Indiana', 'New York', 'Arizona', 'Virginia', 'Tennessee',
       'Alabama', 'South Carolina', 'Oregon', 'Colorado', 'Iowa', 'Ohio',
       'Missouri', 'Oklahoma', 'New Mexico', 'Louisiana', 'Connecticut',
       'New Jersey', 'Massachusetts', 'Georgia', 'Nevada', 'Rhode Island',
       'Mississippi', 'Arkansas', 'Montana', 'New Hampshire', 'Maryland',
       'District of Columbia', 'Kansas', 'Vermont', 'Maine',
       'South Dakota', 'Idaho', 'North Dakota', 'Wyoming',
       'West Virginia'], dtype=object)

In [144]:
df['State'] = df['State'].str.strip()

## 12: PostalCode<a id="12"></a>

In [145]:
# Check if all PostalCode entries are numeric and have 5 digits

df['PostalCode_Check'] = df['PostalCode'].apply(lambda x: x.isdigit() and len(x) == 5)

# Display rows where PostalCode_Check is False

invalid_postal_codes_df = df[~df['PostalCode_Check']]

invalid_postal_codes_df.head(20).T

185  \
PurchaseId                   CA-2016-105018   
PurchaseDate            2016-11-28 00:00:00   
ShipDate                2016-12-02 00:00:00   
ShipMode                     Standard Class   
CustomerId                         SK-19990   
CustomerName                  Sally Knutson   
Segment                            Consumer   
Country                       United States   
City                              Fairfield   
State                           Connecticut   
PostalCode                             6824   
Region                                 East   
ProductId                   OFF-BI-10001890   
Category                    Office Supplies   
SubCategory                         Binders   
ProductName       Avery Poly Binder Pockets   
Sales                                  7.16   
Quantity                                  2   
Discount                                0.0   
Profit                               3.4368   
FirstName                             Sally   
LastName                            Knutson   
PostalCode_Check                      False   

                                                    197                  267  \
PurchaseId                               CA-2017-107720       CA-2016-111010   
PurchaseDate                        2017-11-06 00:00:00  2016-01-22 00:00:00   
ShipDate                            2017-11-13 00:00:00  2016-01-28 00:00:00   
ShipMode                                 Standard Class       Standard Class   
CustomerId                                     VM-21685             PG-18895   
CustomerName                            Valerie Mitchum        Paul Gonzalez   
Segment                                     Home Office             Consumer   
Country                                   United States        United States   
City                                          Westfield           Morristown   
State                                        New Jersey           New Jersey   
PostalCode                                         7090                 7960   
Region                                             East                 East   
ProductId                               OFF-ST-10001414      OFF-FA-10003472   
Category                                Office Supplies      Office Supplies   
SubCategory                                     Storage            Fasteners   
ProductName       Decoflex Hanging Personal Folder File  Bagged Rubber Bands   
Sales                                             46.26                 7.56   
Quantity                                              3                    6   
Discount                                            0.0                  0.0   
Profit                                          12.0276               0.3024   
FirstName                                       Valerie                 Paul   
LastName                                        Mitchum             Gonzalez   
PostalCode_Check                                  False                False   

                                  298  \
PurchaseId             CA-2016-142545   
PurchaseDate      2016-10-28 00:00:00   
ShipDate          2016-11-03 00:00:00   
ShipMode               Standard Class   
CustomerId                   JD-15895   
CustomerName         Jonathan Doherty   
Segment                     Corporate   
Country                 United States   
City                       Belleville   
State                      New Jersey   
PostalCode                       7109   
Region                           East   
ProductId             OFF-PA-10002105   
Category              Office Supplies   
SubCategory                     Paper   
ProductName                 Xerox 223   
Sales                            32.4   
Quantity                            5   
Discount                          0.0   
Profit                         15.552   
FirstName                    Jonathan   
LastName                      Doherty   
PostalCode_Check                False   

     

In [146]:
invalid_postal_codes_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 449 entries, 185 to 9969
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   PurchaseId        449 non-null    object        
 1   PurchaseDate      449 non-null    datetime64[ns]
 2   ShipDate          449 non-null    datetime64[ns]
 3   ShipMode          449 non-null    object        
 4   CustomerId        449 non-null    object        
 5   CustomerName      449 non-null    object        
 6   Segment           449 non-null    object        
 7   Country           449 non-null    object        
 8   City              449 non-null    object        
 9   State             449 non-null    object        
 10  PostalCode        449 non-null    object        
 11  Region            449 non-null    object        
 12  ProductId         449 non-null    object        
 13  Category          449 non-null    object        
 14  SubCategory       449 non-nu

#### There are 449 entries which have 4 digit Postalcode in USA, 
#### standard ZIP codes typically have 5 digits. However, leading zeros are common in certain regions, meaning that a ZIP code like "04356" is indeed valid
#### we will keep ZIP codes as strings in DataFrame to preserve any leading zeros and add 0 to the begining of the codes with 4 digits

In [147]:
# Make all PostalCodes have 5 digits

df['PostalCode'] = df['PostalCode'].str.zfill(5)

In [148]:
# Confirm Changes

df['PostalCode_Check'] = df['PostalCode'].apply(lambda x: x.isdigit() and len(x) == 5)

# Display rows where PostalCode_Check is False

invalid_postal_codes_df = df[~df['PostalCode_Check']]

invalid_postal_codes_df

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, PostalCode_Check]
Index: []

## 13: Region<a id="13"></a>

In [149]:
df['Region'].unique()

array(['South', 'West', 'Central', 'East'], dtype=object)

In [150]:
df['Region'].value_counts()

Region
West       3203
East       2848
Central    2323
South      1620
Name: count, dtype: int64

## 14: ProductId<a id="14"></a>

In [151]:
# ensure that the ProductId column starts with the same three letters of the Category column

df['ProductIdLower'] = df['ProductId'].str.lower()
df['CategoryLower'] = df['Category'].str.lower()

# Extract the first 3 characters from the lowercase versions

df['ProductIdPrefix'] = df['ProductIdLower'].str[:3]
df['CategoryPrefix'] = df['CategoryLower'].str[:3]

# Check if the prefixes match

df['Match'] = df['ProductIdPrefix'] == df['CategoryPrefix']

# Filter rows where ProductId and Category prefixes do not match

mismatched_entries = df[~df['Match']]

mismatched_entries

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, PostalCode_Check, ProductIdLower, CategoryLower, ProductIdPrefix, CategoryPrefix, Match]
Index: []

In [152]:
# Drop unnecessary columns
df = df.drop(columns=['PostalCode_Check', 'ProductIdLower', 'CategoryLower',
                      'ProductIdPrefix', 'CategoryPrefix', 'Match'])



In [153]:
# ensure that the two-letter code after the first hyphen in ProductId matches the first two letters of SubCategory

df['ProductIdLower'] = df['ProductId'].str.lower()
df['SubCategoryLower'] = df['SubCategory'].str.lower()

# Extract the two-letter code after the first hyphen in ProductId

df['ProductIdCode'] = df['ProductIdLower'].apply(lambda x: x.split('-')[1])

# Extract the first two characters of SubCategory

df['SubCategoryCode'] = df['SubCategoryLower'].str[:2]

# Check if the codes match

df['Match'] = df['ProductIdCode'] == df['SubCategoryCode']

# Filter rows where the codes do not match

mismatched_entries = df[~df['Match']]

mismatched_entries

Empty DataFrame
Columns: [PurchaseId, PurchaseDate, ShipDate, ShipMode, CustomerId, CustomerName, Segment, Country, City, State, PostalCode, Region, ProductId, Category, SubCategory, ProductName, Sales, Quantity, Discount, Profit, FirstName, LastName, ProductIdLower, SubCategoryLower, ProductIdCode, SubCategoryCode, Match]
Index: []

In [154]:

df = df.drop(columns=['ProductIdLower', 'SubCategoryLower', 'ProductIdCode',
                      'SubCategoryCode', 'Match'])

df = df.reset_index(drop=True)

## 15: Category<a id="15"></a>